### ISSUE TITLE
How can i get the pretrained_teacher model by myself?

### ISSUE THREAD

**Issue by SteveTanggithub** [RicherMans/PSL#2]
How can i get the pretrained_teacher model by myself instead of using the ones u provided?

**Comment by RicherMans** [comment]
Hey there,
just train a classifier on the standard 10s scale on Audioset.
You might notice, but I generally avoid publishing that code that uses 10s training on Audioset, mainly because of the issue when downloading the dataset, which would lead to a lot of questions.

There are just some caveats that are not really discussed in the paper, but are vital to the success of a good teacher:

1. For CNNs/CRNNs, do not use global average (or any global -) pooling methods. These generally do not generalize to smaller-sized inputs that are necessary for PSL. Instead, always use decision-mean or -max pooling that can accurately predict single time-level audio tags. [Here](https://github.com/fschmid56/EfficientAT/issues/5) is an issue where we discussed the downfalls of global pooling methods.
2. For Transformers, the positional embedding is crucial. Many implementations that I have seen use a wrong positional embedding. For example [AST](https://github.com/YuanGongND/ast/tree/master) uses a time-frequency positional embedding, which is not variable in its size for cases where the input size changes. Thus that repo uses the extreme solution of always padding to 10s, which is completely unbearable in my opinion. The easiest way to avoid any problems is to have time/frequency independent positional embeddings, as proposed in [passt](https://github.com/kkoutini/passt_hear21), which also scale to smaller input lengths. These embeddings make actually more sense than any `traditional` positional embedding of a transformer: In a spectrogram, the position of your frequency is always fixed, you only need to encode that position a single time across all time-frames and not give each position a `unique` representation.

**Comment by zuowanbushiwo** [comment]
Hi @RicherMans 
 
Can your new [CED-Base](https://zenodo.org/record/8275347/files/audiotransformer_base_mAP_4999.pt?download=1)  mAP_4999 model be used as a teacher? From your CED (Consistent ensemble distillation for audio tagging) paper and the corresponding open source [github project](https://github.com/RicherMans/CED), CED-Base is a very powerful model (probably the best publicly available model on AT). And I carefully looked at the use of CED's Transformers, which have time/frequency independent positional embeddings. I think CED-Base can be used as a teacher? Is that right?

Thanks！

**Comment by RicherMans** [comment]
Hey @zuowanbushiwo ,
sure it can be used, actually any model can be used, just any model that supports variable length inference ( AST is here pretty bad, since padding to 10s will lead a lot of silence, that the model probably detects as such).
Kind regards,
Heinrich

**Comment by zuowanbushiwo** [comment]
Hi @RicherMans 

Thanks

**Comment by zuowanbushiwo** [comment]
Hi @RicherMans 

能请问一下吗？我使用 [CED-Base](https://zenodo.org/record/8275347/files/audiotransformer_base_mAP_4999.pt?download=1) 做teacher 模型，但是结果并没有得到提升？ 我是使用 CED-Base 做teacher， 然后使用你的 UIT_Mobile 文章中的 uit_xs 模型做 student， 训练数据集是AudioSet-2M，chunk_len=2s ,但是结果只能到 mAP 是 0.4109。这个正常吗？

谢谢！

**Comment by RicherMans** [comment]
嗯嗯正常，因为模型看不到整个上下文啊。
Audioset 就是有各种问题。
你训练用2s 导致模型的召回会偏高（因为5次推理一条），导致mAP 会低 （mAP = 精度）。

你拿模型看看实际输出是什么，然后对比跟一个10s 模型。

**Comment by zuowanbushiwo** [comment]
<img width="428" height="370" alt="Image" src="https://github.com/user-attachments/assets/dca25270-6e5c-448e-9758-970b3e85e571" />

从这个图片中看我的 Baseline (MA) Full  换成了  [CED-Base](https://zenodo.org/record/8275347/files/audiotransformer_base_mAP_4999.pt?download=1) mAP_4999 的模型，base 模型的 性能提升很多， 而且我用的 student 是 UIT_Mobile  [uit_xs ](https://github.com/RicherMans/UIT_Mobile/blob/973a82661a71ee19bd92748c07602759259ed03d/models/uit.py#L581)

```
def uit_xs(**kwargs):
    model_kwargs = dict(
        patch_size=16,
        embed_dim=128,
        depth=12,
        num_heads=2,
        mlp_ratio=3.0,
        pooling='mean',
        init_bn=True,
        drop_path_rate=0.0,
        act_layer=nn.ReLU,
        attention_type='BNeckAttention',
    )
    model_kwargs.update(
        (k, kwargs[k]) for k in set(kwargs).intersection(model_kwargs))
    model_kwargs = {**model_kwargs, **kwargs}
    return UITBase(**model_kwargs)
```
我以为teacher 和 student 的模型的能力都有很好的提升可以得到更好的结果，但是实际的结果竟然差不多。

>  你训练用2s 导致模型的召回会偏高（因为5次推理一条），导致mAP 会低 （mAP = 精度）。
 你拿模型看看实际输出是什么，然后对比跟一个10s 模型。

这个是什么意思？是因为audioset 本身是10s的数据标签，我们 5次推理一条 计算得到的 mAP 值不能反应模型的能力？

可以用audioset strong的数据集做微调吗？ 能提升模型的能力吗？  audioset strong 有时间标签，如果我们把 audioset strong 按照2s的分割，能得到这2s的实际精确标签？

谢谢！

**Comment by RicherMans** [comment]
> 我以为teacher 和 student 的模型的能力都有很好的提升可以得到更好的结果，但是实际的结果竟然差不多。

嗯嗯，我一开始也认为，但是实际上不是。从LLM 角度了说这个事情很好理解：没有context。好多LLM 性能是好的因为他们支持很长的context ( 几千token),导致可以从context 获得一些信息，帮助解决问题。
你对比两个模型的概念是一样的：2s 模型上下文有限，10s 更长所以可以获得更好的信息。

> 这个是什么意思？是因为audioset 本身是10s的数据标签，我们 5次推理一条 计算得到的 mAP 值不能反应模型的能力？

还是推荐看看模型输出和audioset 标签的关系。audioset-eval 很多标签非常有问题，如(Silence)： https://research.google.com/audioset/eval/silence.html

如果你拿一个合理的模型，那么这个模型应该会在每个2s 有静音，输出静音。但是实际上标签不一定是这样。


> 可以用audioset strong的数据集做微调吗？

这个数据对我来说非常没什么用。微调后模型输出都是mechanisms, generic impact sounds 。很久以前试图过用这个数据，结论就是完全不如Audioset weak。

>  能提升模型的能力吗？ audioset strong 有时间标签，如果我们把 audioset strong 按照2s的分割，能得到这2s的实际精确标签？

如果你用 Audioset Strong 测试集，应该会有提升。在这个论文测试过： https://arxiv.org/pdf/2305.17834 。

总结：提升不大是因为测试数据 + 指标有一些问题。

**Comment by zuowanbushiwo** [comment]
好的，非常感谢您的耐心回复，我再仔细思考你上面的话。
